MantidPlot/src/Mantid/MantidDock.cpp

```diff
@@ -1517,53 +1524,245 @@ void MantidDockWidget::plotSurface() {
+  const QStringList wsNames = m_tree->getSelectedWorkspaceNames();
+  if (!wsNames.empty()) {
+    const auto wsName = wsNames[0];
+    const auto wsGroup = boost::dynamic_pointer_cast<const WorkspaceGroup>(
+        m_ads.retrieve(wsName.toStdString()));
+    if (wsGroup) {
+      auto options = m_tree->chooseContourPlotOptions();
+      // If user clicked OK and not Cancel...
+      if (options.accepted) {
+
+        // Set up one new matrix workspace to hold all the data for plotting
+        const QString plotWSTitle("__matrixToPlotContour");
+        const QString xLabelQ =
+            createWorkspaceForGroupPlot(plotWSTitle, wsGroup, options);
+
+        // Convert to histogram data if not already
+        convertPointsToHisto(plotWSTitle);
+
+        // Plot the output workspace as a contour plot
+        auto matrixToPlot = m_mantidUI->importMatrixWorkspace(plotWSTitle, -1,
+                                                              -1, false, false);
+        m_mantidUI->deleteWorkspace(plotWSTitle);
+
+        // Copying the plot in this way ensures that the workspace can be
+        // deleted and the plot will not disappear
+        MultiLayer *plot = m_appParent->multilayerPlot(
+            m_appParent->generateUniqueName(tr("Graph")));
+        plot->removeLayer();
+        plot->copy(matrixToPlot->plotGraph2D(Graph::ColorMapContour));
+
+        // Set the X, Y axis labels correctly
+        plot->activeGraph()->setXAxisTitle(xLabelQ);
+        plot->activeGraph()->setYAxisTitle(options.axisName);
+
+        // Default title is "__matrixToPlotContour". Change this:
+        QString title = QString("Contour plot for %1, spectrum %2")
+                            .arg(wsGroup->name().c_str(),
+                                 QString::number(options.plotIndex));
+        plot->activeGraph()->setTitle(title);
+      }
+    }
+  }
+}
+
+/**
+ * Create a workspace for the surface/contour plot from the given workspace
+ * group, and add it to the ADS. Returns title for the X axis.
+ * Called by plotSurface() and plotContour().
+ * @param wsName :: [input] Name to give the resulting workspace
+ * @param wsGroup :: [input] Pointer to workspace group to use as input
+ * @param options :: [input] User input from dialog
+ * @returns Title for the X axis, read from the workspaces in the group
+ */
+const QString MantidDockWidget::createWorkspaceForGroupPlot(
+    const QString &wsName, WorkspaceGroup_const_sptr wsGroup,
+    const MantidSurfacePlotDialog::UserInputSurface &options) {
+  QString xAxisTitle;
+  int index = options.plotIndex; // which spectrum to plot from each WS
+  QString logName = options.logName; // Log to read from for axis of XYZ plot
+  if (wsGroup) {
+    // Create workspace to hold the data
+    // Each "spectrum" will be the data from one workspace
+    int nWorkspaces = wsGroup->getNumberOfEntries();
+    if (nWorkspaces > 0) {
+      const auto firstWS = boost::dynamic_pointer_cast<const MatrixWorkspace>(
```

Comment:
How are you protecting about failure to downcast in the dynamic_pointer_cast. We'll get a nullptr if the group contained table or peaks workspaces for example.